Mathlib/Tactic/Algebraize.lean

```diff
@@ -0,0 +1,262 @@
+      if ¬ (← t.anyM (fun j => Meta.isDefEq j f)) then return
+
+      let cinfo ← getConstInfo p
+      let n ← getExpectedNumArgs cinfo.type
+      let pargs := Array.mkArray n (none : Option Expr)
+      /- If the attribute points to the corresponding `Algebra` property itself, we assume that it
+      is definitionally the same as the `RingHom` property. Then, we just need to construct its type
+      and the local declaration will already give a valid term. -/
+      if cinfo.isInductive then
+        let pargs := pargs.set! 0 args[0]!
+        let pargs := pargs.set! 1 args[1]!
+        let tp ← mkAppOptM p pargs -- This should be the type `Algebra.Property A B`
+        liftMetaTactic fun mvarid => do
+          let nm ← mkFreshBinderNameForTactic `AlgebraizeInst
+          let (_, mvar) ← mvarid.note nm decl.toExpr tp
+          return [mvar]
+      /- Otherwise, the attribute points to a constructor of the `Algebra` property. In this case,
+      we assume that the `RingHom` property is the last argument of the constructor (and that
+      this is all we need to supply explicitly). -/
+      else
+        let pargs := pargs.set! (n - 1) decl.toExpr
+        let val ← mkAppOptM p pargs
+        liftMetaTactic fun mvarid => do
+          let nm ← mkFreshBinderNameForTactic `AlgebraizeInst
+          let (_, mvar) ← mvarid.note nm val
+          return [mvar]
+    | none => return
+
+end Algebraize
+
+open Algebraize
+
+/-- Tactic that, given `RingHom`s, adds the corresponding `Algebra` and (if possible)
+`IsScalarTower` instances, as well as `Algebra` corresponding to `RingHom` properties available
+as hypotheses.
+
+Example: given `f : A →+* B` and `g : B →+* C`, and `hf : f.FiniteType`, `algebraize f g` will add
+the instances `Algebra A B`, `Algebra B C`, and `Algebra.FiniteType A B`.
+
+See the `algebraize` tag for instructions on what properties can be added. -/
+syntax "algebraize" (ppSpace colGt term:max)* : tactic
+
+elab_rules : tactic
+  | `(tactic| algebraize $[$t:term]*) => do
+    let t ← t.mapM fun i => Term.elabTerm i none
+    -- We loop through the given terms and add algebra instances
+    for f in t do
+      let ft ← inferType f
+      match ft.getAppFn with
+      | Expr.const ``RingHom _ => addAlgebraInstanceFromRingHom f ft
+      | _ => throwError m!"{f} is not of type `RingHom`"
+    -- After having added the algebra instances we try to add scalar tower instances
+    for f in t do
+      match f.getAppFn with
+      | Expr.const ``RingHom.comp _ =>
+        try addIsScalarTowerInstanceFromRingHomComp f
+        catch _ => continue
+      | _ => continue
+
+    -- Search through the local context to find other instances of algebraize
+    searchContext t
+
+/-- Version of `algebraize`, which only adds `Algebra` instances and `IsScalarTower` instances. -/
+syntax "algebraize'" (ppSpace colGt term:max)* : tactic
+
+elab_rules : tactic
+  | `(tactic| algebraize' $[$t:term]*) => do
```

Comment:
I think after fixing a few stylistic things this would be ready to merge. Thank you for the cool contribution!

One final question, since you end up using `algebraize'` quite often, I wonder if the two versions should be named `algebraize!` and `algebraize` (instead of `algebraize` and `algebraize'`).
Maybe that could be a question for Zulip or for @adamtopaz / @jcommelin?

Reply:
I think that's a good idea, but I can ask in Zulip also.

Reply:
Could you please explain the difference between the two in a bit more detail?

Reply:
@jcommelin the only difference is that `algebraize'` does not try to add algebra properties corresponding to ringhom properties that are available as hypothesis. I noticed that there were some statements where such properties where available, but they didn't need to be algebraized, and of course when there are no ringhom properties available one does not need to search for them.

I guess peformance wise it might be a neglegible improvement, but I still thought it seemed unecessary to peform this search when it was not needed.

Reply:
and thank you @joneugster for the review & suggestions!

Reply:
I see. I do wonder whether it's worth it from a UX point of view.
Would it be possible to turn this into some boolean flag that can be passed as config option to `algebraize`?

I'm also fine with having them both as is, and then we can evaluate in a few months to see whether `algebraize'` is used much in practice.

Reply:
I actually like the config idea. Here is a snippet how other tactics like `simp`, `rw`, etc. do this:

```lean
namespace Algebraize

[...]

structure Config where
  searchContext : Bool := true
deriving Inhabited

declare_config_elab elabAlgebraizeConfig Algebraize.Config

end Algebraize

open Algebraize Lean.Parser.Tactic

/-- Tactic that, given `RingHom`s, adds the corresponding `Algebra` and (if possible)
`IsScalarTower` instances, as well as `Algebra` corresponding to `RingHom` properties available
as hypotheses.

Example: given `f : A →+* B` and `g : B →+* C`, and `hf : f.FiniteType`, `algebraize f g` will add
the instances `Algebra A B`, `Algebra B C`, and `Algebra.FiniteType A B`.

See the `algebraize` tag for instructions on what properties can be added. -/
syntax "algebraize" (config)? (ppSpace colGt term:max)* : tactic

elab_rules : tactic
  | `(tactic| algebraize $[$config]? $[$t:term]*) => do
    let cfg ← match config with
      | some conf => elabAlgebraizeConfig conf
      | none => pure default
    let t ← t.mapM fun i => Term.elabTerm i none
    -- We loop through the given terms and add algebra instances
    for f in t do
      let ft ← inferType f
      match ft.getAppFn with
      | Expr.const ``RingHom _ => addAlgebraInstanceFromRingHom f ft
      | _ => throwError m!"{f} is not of type `RingHom`"
    -- After having added the algebra instances we try to add scalar tower instances
    for f in t do
      match f.getAppFn with
      | Expr.const ``RingHom.comp _ =>
        try addIsScalarTowerInstanceFromRingHomComp f
        catch _ => continue
      | _ => continue

    -- Search through the local context to find other instances of algebraize
    if cfg.searchContext then
      searchContext t

```

and then it's used with `algebraize (config := { searchContext := false })` (or a less verbose version thereof). Maybe you can think about what the best name for the config option is.

Reply:
> Maybe what I can do, is use the config setup that you gave above, and then define algebraize' to be just a macro / abbrevation for algebraize (config := { searchContext := false })?

Let's to that, that sounds like a good suggestion!